File: server/src/test/kotlin/LifecycleTest.kt
package ch.nokillswit

import ch.nokillswit.authz.ConflictException
import ch.nokillswit.contracts.Lifecycle
import ch.nokillswit.contracts.canTransitionTo
import ch.nokillswit.contracts.contentEditable
import ch.nokillswit.contracts.deletable
import ch.nokillswit.contracts.published
import ch.nokillswit.contracts.requireTransitionTo
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/** The strict lifecycle state machine — pinned as a whole matrix so no edge slips in unnoticed. */
class LifecycleTest {

    @Test
    fun `the transition matrix is exactly the documented one`() {
        val allowed = Lifecycle.entries.flatMap { from -> Lifecycle.entries.filter { from.canTransitionTo(it) }.map { from to it } }
        assertEquals(
            setOf(
                Lifecycle.DRAFT to Lifecycle.PROPOSED,
                Lifecycle.PROPOSED to Lifecycle.DRAFT,
                Lifecycle.PROPOSED to Lifecycle.ACTIVE,
                Lifecycle.ACTIVE to Lifecycle.DEPRECATED,
                Lifecycle.DEPRECATED to Lifecycle.RETIRED,
            ),
            allowed.toSet(),
        )
        assertFailsWith<ConflictException> { Lifecycle.DRAFT.requireTransitionTo(Lifecycle.ACTIVE) }
        assertFailsWith<ConflictException> { Lifecycle.RETIRED.requireTransitionTo(Lifecycle.DRAFT) }
        Lifecycle.DRAFT.requireTransitionTo(Lifecycle.PROPOSED)
    }

    @Test
    fun `content is editable only before publication, only a draft is deletable`() {
        assertTrue(Lifecycle.DRAFT.contentEditable)
        assertTrue(Lifecycle.PROPOSED.contentEditable)
        for (l in listOf(Lifecycle.ACTIVE, Lifecycle.DEPRECATED, Lifecycle.RETIRED)) assertFalse(l.contentEditable, "$l")
        assertTrue(Lifecycle.DRAFT.deletable)
        assertFalse(Lifecycle.PROPOSED.deletable)
        assertTrue(Lifecycle.ACTIVE.published)
        assertTrue(Lifecycle.DEPRECATED.published)
        assertFalse(Lifecycle.RETIRED.published)
    }
}

File: web/src/utils/lifecycle.test.ts
import { describe, expect, test } from "vitest";
import { isContentEditable, isDeletable, isIrreversibleTransition, LIFECYCLES, TRANSITIONS } from "./lifecycle";

describe("lifecycle", () => {
  test("the matrix mirrors the server: forward only, PROPOSED may step back, RETIRED is terminal", () => {
    expect(TRANSITIONS.DRAFT).toEqual(["PROPOSED"]);
    expect(TRANSITIONS.PROPOSED).toEqual(["DRAFT", "ACTIVE"]);
    expect(TRANSITIONS.ACTIVE).toEqual(["DEPRECATED"]);
    expect(TRANSITIONS.DEPRECATED).toEqual(["RETIRED"]);
    expect(TRANSITIONS.RETIRED).toEqual([]);
    expect(LIFECYCLES).toHaveLength(5);
  });

  test("editable and deletable follow the publication line", () => {
    expect(LIFECYCLES.filter(isContentEditable)).toEqual(["DRAFT", "PROPOSED"]);
    expect(LIFECYCLES.filter(isDeletable)).toEqual(["DRAFT"]);
    expect(LIFECYCLES.filter(isIrreversibleTransition)).toEqual(["DEPRECATED", "RETIRED"]);
  });
});
